call.go

```diff
@@ -31,12 +31,30 @@ func (cc *ClientConn) Invoke(ctx context.Context, method string, args, reply int
 	// configured as defaults from dial option as well as per-call options
 	opts = combine(cc.dopts.callOptions, opts)
 
+	interceptors := cc.dopts.chainUnaryInts
+
+	// Prepend dopts.unaryInt to the chaining interceptors if it exists, since unaryInt will
+	// be executed before any other chained interceptors.
 	if cc.dopts.unaryInt != nil {
-		return cc.dopts.unaryInt(ctx, method, args, reply, cc, invoke, opts...)
+		interceptors = append([]UnaryClientInterceptor{cc.dopts.unaryInt}, interceptors...)
+	}
+
+	if len(interceptors) > 0 {
+		return interceptors[0](ctx, method, args, reply, cc, getChainInvoker(interceptors, 0, invoke), opts...)
 	}
 	return invoke(ctx, method, args, reply, cc, opts...)
 }
 
+// getChainInvoker recursively generate the chained unary invoker.
+func getChainInvoker(interceptors []UnaryClientInterceptor, curr int, finalInvoker UnaryInvoker) UnaryInvoker {
```

Comment:
This works but looks like an overkill. Recursion like this is hard to understand and maintain.

This function builds the chain top-down. It's only necessary when you don't have the real `invoke`.
Since we have `invoke`, we should be able to build the chain bottom-up.

```go
func chain(its []UnaryClientInterceptor, finalInvoker UnaryInvoker) UnaryInvoker {
	ret := finalInvoker
	for i := len(its) - 1; i >= 0; i-- {
		it, inv := its[i], ret // Trap values in local variables.
		ret = func(ctx context.Context, method string, req, reply interface{}, cc *ClientConn, opts ...CallOption) error {
			return it(ctx, method, req, reply, cc, inv, opts...)
		}
	}
	return ret
}
```

It also works with 0 interceptors, so in `ClientConn.Invoke`, you could just do 
```go
chain(interceptors, invoke)(ctx, method, args, reply, cc, opts...)
```

Reply:
Nice. I have adapted your logic.